packages/ckeditor5-html-support/tests/htmlcomment.js

```diff
@@ -0,0 +1,184 @@
+		it( 'should not create a dedicated model element for a comment node', () => {
+			editor.setData( '<p><!-- comment 1 -->Foo<!-- comment 2 --></p>' );
+
+			expect( getModelData( model, { withoutSelection: true } ) ).to.equal( '<paragraph>Foo</paragraph>' );
+		} );
+	} );
+
+	describe( 'data downcast conversion', () => {
+		it( 'should convert each $comment marker to a comment node', () => {
+			editor.setData( '<p>Foo</p>' );
+
+			const paragraph = root.getChild( 0 );
+
+			addMarker( '$comment:1', ' comment 1 ', paragraph, 1 );
+			addMarker( '$comment:2', ' comment 2 ', paragraph, 2 );
+
+			expect( editor.getData() ).to.equal( '<p>F<!-- comment 1 -->o<!-- comment 2 -->o</p>' );
+		} );
+
+		it( 'should convert each $comment marker to a comment node inside a nested tree', () => {
+			editor.setData( '<div><div><div><p>Foo</p></div></div></div>' );
+
+			const paragraph = root.getNodeByPath( [ 0, 0, 0, 0 ] );
+
+			addMarker( '$comment:1', ' comment 1 ', paragraph, 1 );
+			addMarker( '$comment:2', ' comment 2 ', paragraph, 2 );
+
+			expect( editor.getData() ).to.equal( '<div><div><div><p>F<!-- comment 1 -->o<!-- comment 2 -->o</p></div></div></div>' );
+		} );
+	} );
+
+	describe( 'marker removal post fixer', () => {
+		it( 'should remove all markers and corresponding $root attributes', () => {
+			editor.setData( '<p><!-- comment 1 -->Foo<!-- comment 2 --></p>' );
+
+			model.change( writer => {
+				writer.remove( writer.createRangeIn( root ) );
+			} );
+
+			const rootAttributes = [ ...root.getAttributeKeys() ].filter( attr => attr.startsWith( '$comment' ) );
+
+			expect( rootAttributes ).to.have.length( 0 );
+
+			const markers = [ ...editor.model.markers ].filter( marker => marker.name.startsWith( '$comment' ) );
+
+			expect( markers ).to.have.length( 0 );
+		} );
+
+		it( 'should remove markers and corresponding $root attributes representing removed comments only', () => {
+			editor.setData( '<p><!-- comment 1 -->Foo<!-- comment 2 --></p><p><!-- comment 3 -->Foo<!-- comment 4 --></p>' );
+
+			model.change( writer => {
+				const firstParagraph = root.getChild( 0 );
+
+				writer.remove( writer.createRangeOn( firstParagraph ) );
+			} );
+
+			const rootAttributes = [ ...root.getAttributeKeys() ].filter( attr => attr.startsWith( '$comment' ) );
+
+			expect( rootAttributes ).to.have.length( 2 );
+
+			const markers = [ ...editor.model.markers ].filter( marker => marker.name.startsWith( '$comment' ) );
+
+			expect( markers ).to.have.length( 2 );
+
+			expect( root.getAttribute( markers[ 0 ].name ) ).to.equal( ' comment 3 ' );
+			expect( root.getAttribute( markers[ 1 ].name ) ).to.equal( ' comment 4 ' );
+		} );
+	} );
+
+	function addMarker( name, content, element, offset ) {
+		model.change( writer => {
+			writer.addMarker( name, {
+				usingOperation: true,
+				affectsData: true,
+				range: writer.createRange(
+					writer.createPositionAt( element, offset )
+				)
+			} );
+
+			writer.setAttribute( name, content, root );
```

Comment:
Could we wrap this in a separate method? It's not a part of adding markers.

Reply:
Done.
I added symmetrical methods: `_getCommentContent( markerName )` and `_setCommentContent( markerName, commentContent )`.

Reply:
I got it wrong. The private methods mentioned in my comment above are not needed (and now are removed). Changes were needed in test file, where the `addMarker()` function actually was doing 2 things: adding a marker and setting a root attribute.

Now `addMarker()` adds a marker only, and root attribute is added separately.